File: python/langchain-basic-agent/src/tools.py
"""LangChain tool definitions for the langchain-basic-agent example.

Three tools are defined:
  - get_weather  — safe information tool (ALLOWED by policy)
  - delete_files — destructive file tool (DENIED by policy)
  - send_email   — outbound communication tool (requires APPROVAL)
"""
from __future__ import annotations

from langchain_core.tools import tool


@tool
def get_weather(city: str) -> str:
    """Get the current weather for a city.

    Returns mock data in offline / demo mode.
    """
    return f"🌤  Weather in {city}: 22°C, partly cloudy (mock response)"


@tool
def delete_files(path: str) -> str:
    """Delete all files at the given filesystem path.

    This tool is DANGEROUS and is blocked by policy in this example.
    """
    return f"Deleted files at {path}"


@tool
def send_email(to: str, subject: str, body: str) -> str:
    """Send an email to the specified recipient.

    This tool requires human approval before execution in this example.
    """
    return f"Email sent to {to} — subject: {subject}"

File: node/langchain-js-basic-agent/src/tools.ts
export interface ToolResult {
  output: string;
  allowed: boolean;
}

export function getWeather(location: string): ToolResult {
  return {
    output: `Weather in ${location}: 22°C, partly cloudy. [mock]`,
    allowed: true,
  };
}

export function deleteFile(path: string): ToolResult {
  return {
    output: `[BLOCKED] deleteFile("${path}") denied by policy.`,
    allowed: false,
  };
}

export const TOOLS = {
  get_weather: (args: Record<string, unknown>): ToolResult =>
    getWeather(typeof args["location"] === "string" ? args["location"] : "unknown"),
  delete_file: (args: Record<string, unknown>): ToolResult =>
    deleteFile(typeof args["path"] === "string" ? args["path"] : ""),
} as const;

export type ToolName = keyof typeof TOOLS;
